daemon/cdi.go

```diff
@@ -0,0 +1,44 @@
+package daemon
+
+import (
+	"fmt"
+
+	"github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	"github.com/docker/docker/errdefs"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// initializeCdi initializes the CDI registry and registers the CDI device driver.
+func initializeCdi(d *Daemon) error {
+	if !d.HasExperimental() {
+		return errdefs.NotImplemented(errors.New("CDI is only supported in experimental mode"))
+	}
+
+	return registerCdiDriver()
+}
```

Comment:
This is no longer conditional, so it'll always fail startup in non-experimental mode, and on Windows even in experimental mode.

In this flow, the check probably needs to move down into `injectDevices` before `cdi.GetRegistry()` (and `registerCdiDriver` goes back to being a no-op on non-Linux), which probably means the failure goes back to InvalidArgument, like the checkpoint example you linked earlier.

That way the CDI driver always exists (on LInux) but fails if you try to use it in non-experimental mode.

I'm not a Docker maintainer, but _in my opinion_ there's no need to gate this feature, since it only has effect if you actually send a CDI DeviceRequest to the API; so its presence is pretty non-risky, particularly now that it lazy-initialises the registry.

That said, if there's some consideration that the API or exposure might change in the future (after this makes it into a Docker release), experimental still makes sense, to reflect that.

Reply:
Yes, of course. I missed this while trying to simplify things. Need to take another pass at it then.

Reply:
What I am thinking is that I will convert `registerCdiDriver` to a no-op on non-linux platforms. Since no device driver is ever registered this will either trigger an error if a `DeviceRequest` which should match `cdi` is ever detected, or never be triggered.

With regards to gating this behind the experimental flag. This was at the request of the maintainers (/cc @corhere) and as such I will leave it in for this interation.

I do agree that with the lazy instantiation of the registry the code is only ever triggered when a user actually reqested a device (on a supported platform) and as such the risks are reduced.

Reply:
> there's no need to gate this feature, since it only has effect if you actually send a CDI DeviceRequest to the API

The risk we want to mitigate by gating it behind a feature flag is [the risk that people will start to depend on it](https://www.hyrumslaw.com) before the feature has been stabilized, forcing us to support the ostensibly-experimental API forever.

Reply:
That was kind-of my point: There's not a lot I can envisage changing about this feature (since we're just passing a device string in a known format into an implementation library), and particularly, not a lot I can see that would need to change before removing the experimental flag, i.e. I can't see what the exit-condition here is, except "It's been in the codebase for a while", and unless the Docker release cycle is wildly accelerating, that condition seems (from the outside) pretty close to "Got into a stable release".

Putting this behind experimental (rather than a specific feature flag) actually _invokes_ Hyrum's rule, because anyone who wants to use CDI (which is a deliberate choice at container-run-time anyway, you can't invoke it by accident) has now exposed _all users_ of that daemon instance to also relying on other experimental features they may not realise are experimental. I think experimental makes more sense for behaviour _changes_, i.e. already-valid things that may do something differently, and are intended to be true in the future.

I also tend to assume that anyone running a non-released build from, e.g., nightlies, is expecting to be exposed to experimental features, but that's probably an over-reach: on reflection, this was a large part of my rationale for suggesting not gating this at all, and that is probably a misalignment on my part.

Anyway, maybe the resolution I'd like to see here is to explicitly document the _experiment_: What's the condition that will let us remove the experimental gate on this feature, even if it's "Made it into a release and none of the `--experimental` users reported issues" or similar. And I apologise if there's already a project-wide definition for that which I'm unaware of.

Reply:
I have created https://github.com/moby/moby/issues/45192 to track what is required to remove the `--experimental` gating here. Once things have settled in this PR, I can add additional tasks and context there.